### ISSUE TITLE
range_zip_with_len causes type mismatch

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#14929]
### Summary

.

### Lint Name

range_zip_with_len

### Reproducer

I tried this code:

```rust
pub fn iter_too(v: &[u32]){
    for (_, i) in v.iter().zip(0..v.len()) {
        assert!(i <= v.len());
    }
}

pub fn main() {}

```

I saw this happen:

```
warning: using `.zip()` with a range and `.len()`
 --> src/main.rs:2:19
  |
2 |     for (_, i) in v.iter().zip(0..v.len()) {
  |                   ^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `v.iter().enumerate()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#range_zip_with_len
  = note: `#[warn(clippy::range_zip_with_len)]` on by default
```

The suggested code

```rust
pub fn iter_too(v: &[u32]){
    for (_, i) in v.iter().enumerate() {
        assert!(i <= v.len());
    }
}

pub fn main() {}

```

does not compile
```
error[E0308]: mismatched types
 --> src/main.rs:3:22
  |
3 |         assert!(i <= v.len());
  |                      ^^^^^^^ expected `&u32`, found `usize`

For more information about this error, try `rustc --explain E0308`.
```

### Version

```text

```

**Comment by matthiaskrgr** [comment]
yeh, maybe the lint shouldn't be MachineApplicable in this case :) 
https://rust-lang.github.io/rust-clippy/master/index.html#range_zip_with_len
